### ISSUE TITLE
SDDM on Wayland loops because it needs tty1 to start - but there's getty there.

### ISSUE THREAD

**Issue by NyaomiDEV** [sddm/sddm#1844]
On Artix Linux, tty1 to 6 are `getty` sessions. SDDM (Wayland + KWin) could start on tty7 which is free to use, but there's no configuration file to make it do so.

Also, it will keep trying to get VT 1 regardless! (Of course, `getty` sessions are still on `login` so logind doesn't know about them)

Relevant log (copied from `bat`):
```
12675   │ [12:11:28.113] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/sugar-candy/theme.conf"
12676   │ [12:11:28.114] (II) DAEMON: Greeter starting...
12677   │ [12:11:28.123] (II) HELPER: [PAM] Starting...
12678   │ [12:11:28.123] (II) HELPER: [PAM] Authenticating...
12679   │ [12:11:28.123] (II) HELPER: [PAM] returning.
12680   │ [12:11:28.154] (EE) HELPER: Failed to take control of "/dev/tty1" ("root"): Operation not permitted
12681   │ [12:11:28.154] (II) DAEMON: Greeter session started successfully
12682   │ [12:11:28.155] (II) HELPER: [PAM] Closing session
12683   │ [12:11:28.155] (II) HELPER: [PAM] Ended.
12684   │ [12:11:28.156] (WW) DAEMON: Auth: sddm-helper exited with 5
12685   │ [12:11:28.156] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
12686   │ [12:11:28.156] (II) DAEMON: Jumping to VT 1
12687   │ [12:11:28.185] (II) DAEMON: VT mode didn't need to be fixed
12688   │ [12:11:28.185] (II) DAEMON: Socket server stopping...
12689   │ [12:11:28.186] (II) DAEMON: Socket server stopped.
12690   │ [12:11:28.186] (II) DAEMON: Removing display SDDM::Display(0x55e813be6a10) ...
12691   │ [12:11:28.186] (II) DAEMON: Adding new display...
12692   │ [12:11:28.186] (II) DAEMON: Loaded empty theme configuration
12693   │ [12:11:28.186] (II) DAEMON: Using VT 1
12694   │ [12:11:28.186] (II) DAEMON: Display server started.
12695   │ [12:11:28.186] (II) DAEMON: Socket server starting...
12696   │ [12:11:28.186] (II) DAEMON: Socket server started.
12697   │ [12:11:28.186] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/sugar-candy/theme.conf"
12698   │ [12:11:28.187] (II) DAEMON: Greeter starting...
12699   │ [12:11:28.195] (II) HELPER: [PAM] Starting...
12700   │ [12:11:28.195] (II) HELPER: [PAM] Authenticating...
12701   │ [12:11:28.195] (II) HELPER: [PAM] returning.
12702   │ [12:11:28.237] (EE) HELPER: Failed to take control of "/dev/tty1" ("root"): Operation not permitted
```

Ideally, the `SDDM_INITIAL_VT` constant should be picked up from a configuration file, so that we can set that one to 2 and disable `getty` on there. Or there should be a more reliable way to see if a VT is already claimed.

**Comment by Vogtinator** [comment]
SDDM already uses VT_OPENQRY to find out which VT is free. What might be happening in your case is that getty is started on demand, after SDDM switches to it. This shouldn't happen though as normally the tty is opened and taken ownership of before switching to it. Getty would also keep running and the next sddm greeter iteration should pick a different VT. So I'm not sure what's going on.

**Comment by NyaomiDEV** [comment]
In my system `getty` is not started on demand. I can only think that the code fails before the VT check you're talking about.

Also I should note that only root Xorg works and it picks up VT 2 correctly, basically overriding(?) `getty`.

**Comment by NyaomiDEV** [comment]
What I think is happening here is that because of those lines here:

https://github.com/sddm/sddm/blob/501129294be1487f753482c29949fc1c19ef340e/src/daemon/Display.cpp#L127-L128

https://github.com/sddm/sddm/blob/501129294be1487f753482c29949fc1c19ef340e/src/daemon/Display.cpp#L79-L82

https://github.com/sddm/sddm/blob/501129294be1487f753482c29949fc1c19ef340e/src/daemon/Display.cpp#L60-L77

`logind` doesn't know that VTs 1 to 6 are used, because of course there are no users logged in yet, but `getty` is there as a process, and it is using those VTs and also it's being owned by the `root` user. SDDM, running as its own `sddm` user, of course gets a permission error if it tries to get a VT which is used by a `root` owned process, and it should try the next one.

Except, it doesn't.

It doesn't because, as I already attached, it asks `logind` to check if the VT is available. The `VT_OPENQRY` ioctl would never happen as the `fetchAvailableVt` function is stuck in a loop of returning `1` because of that condition I showed earlier, and `VirtualTerminal::setUpNewVt()` which actually does have that ioctl in it is only called by `fetchAvailableVt` if there was no `logind` at all.

**Comment by Vogtinator** [comment]
https://github.com/sddm/sddm/pull/1896 might fix this

**Comment by fxzxmicah** [comment]
Temporary solution: Add `console=ttyS1` to the kernel parameter and accept the systemd-vconsole-setup service fails.

**Comment by SalahAdDin** [comment]
> Temporary solution: Add `console=ttyS1` to the kernel parameter and accept the systemd-vconsole-setup service fails.

where to put this?

**Comment by Vogtinator** [comment]
Please check whether `DisplayServer=x11-user` works.

**Comment by SalahAdDin** [comment]
> Please check whether `DisplayServer=x11-user` works.

how to do so?

When I launch `startplasma-wayland` from `tty` I have my second monitor working well. But when launching Wayland KDE from SDDM, the second screen is not detected.

**Comment by Vogtinator** [comment]
It's possible I'm mixing up two cases which have different causes but the same error:
a) `DisplayServer=wayland` or `DisplayServer=x11-user` not starting, should be fixed by #1896 
b) `DisplayServer=x11` (the default) + autologin into wayland not starting, should be fixed by #1969

There are various tickets which mix both... Which case is yours?

**Comment by ghost** [comment]
PCLinuxOS also uses getty on 1-6. The current version in the repos is 0.19 which wants SDDM to start on VT8 (using the old MinimumVT configuration option), but regardless of me removing this and setting it with -DSDDM_INITIAL_VT instead on VT7 or 8 with SDDM 0.21 it does not start. I cannot get any explanation from sddm.log. The greeter works via --test-mode but is not launched by sddm.

**Comment by Vogtinator** [comment]
`SDDM_INITIAL_VT` doesn't work, see https://github.com/sddm/sddm/issues/1650

Currently I don't actually see a way to reinstate that properly, as there is no way to check whether a tty is currently in use. The way systemd manages getty@ units means that they're only actually opened when something switches to it.

Old sddm just didn't really care and used whatever VT was next and failed if that didn't work.

**Comment by mrmazda** [comment]
It is allowed to start gettys on VTs on boot as in the past, rather than as normally done with systemd now, on demand. I've been doing it for years, as I have my own routine uses for VTs 1-3 and expect X to appear on 7+ as has been the tradition.

**Comment by Vogtinator** [comment]
What could be done is to look at `VT_GETSTATE`'s `v_state` bitmask and only use whatever is marked as free, only looking at those above the minimum. That means it can't go past tty16 and I'm not aware of any such API for FreeBSD either.

**Comment by NyaomiDEV** [comment]
For now, to get Wayland running (it's kind of important when you have a 4k display) I blacklisted agetty on tty1. This would be enough to make SDDM start, but unfortunately for me I also have Plymouth set up, so I came up with a simple init service I called "sddm-pre" which is very simple:

```bash
#!/usr/bin/openrc-run

depend() {
    before sddm
}

start() {
    ebegin "Telling plymouth to quit before starting SDDM"
    if /usr/bin/plymouth --ping; then
        /usr/bin/plymouth quit
    fi
    eend
}
```

This closes Plymouth and leaves tty1 up for grabs, so SDDM can finally grab it.

**Comment by Vogtinator** [comment]
sddm.service already has `After=plymouth-quit.service`. For openrc I guess some equivalent change might be necessary?

**Comment by NyaomiDEV** [comment]
> For openrc I guess some equivalent change might be necessary?

There's this `openrc-plymouth-plugin` which closes Plymouth as soon as all the services are brought up. Unfortunately, this also means that SDDM is brought up in that meantime, so I needed to explicitly terminate it before SDDM started.

This plugin doesn't really provide a `plymouth-quit` "target" though.

This could probably also be solved by that PR I saw the other day, which handles smooth transitions between it and SDDM. What was it again, #1932  iirc?

**Comment by findurance** [comment]
SDDM version 0.21

My SDDM has been doing the same. 1st login just returns to the SDDM greeter.

Today, it jumped into VT1, and stayed there. ON VT2 SDDM greeter was in "inactive" view, with password entry filled. Had to `sudo service sddm restart` out of it.

This renders SDDM totally unusable!
[opensuse-sddm-login-ends-up-in-vt1.txt](https://github.com/user-attachments/files/17173624/opensuse-sddm-login-ends-up-in-vt1.txt)

**Comment by v-evets** [comment]
IMO this free VT detection experiment has been dragging on for far too long now. It's no more reliable now than when it first landed and it's breaking more important features like x11-user and wayland support. 
A suitable wayland display manager is currently the only thing holding up a complete X11-free KDE Plasma configuration, and if it weren't for this particular aggravation SDDM would be *the* answer.

Can't we just put the dumb "try VTs starting from INITIAL_VT, else fail" behaviour back (or better yet, a simple static VT assignment override in the config file) for a while, and hash this automation out in a dev branch or something? Pretty please?

SDDM dying if its configured VT is taken is a simple configuration issue and easy to diagnose for the end-user. 
SDDM launching on random TTYs, going into endless loops, or otherwise behaving eratically and unpredictably because "that's how systemd says this should work" (but it doesn't) is not progress.

**Comment by ghost** [comment]
On Artix + Dinit, disabled service tty1 

and in /etc/dinit.d/config/console.conf set

`ACTIVE_CONSOLES="/dev/tty[2-6]"
`

SDDM starts with `DisplayServer=wayland`

**Comment by Megver83** [comment]
Disabling tty1 did also work for me (Parabola GNU/Linux-libre, OpenRC, seatd)

```
# rc-update del agetty.tty1
```

Tested with wayland and x11 display servers

**Comment by agowa** [comment]
I managed to nail down the root cause of this issue.

It's a wird edge case interaction. Thanks goes to @Vogtinator who provided enough information to get starting debugging with especially this comment here https://github.com/sddm/sddm/pull/1896#issuecomment-2004263424
> Apparently `getty@.service` has `Type=idle` which means it is marked as active before anything actually happened. I don't see a way to fix this.

(The entire journey of trial and error so far: https://chaos.social/@agowa338/114853300563852266 )

Tl;Dr: sddm should use the TTY capabilities of systemd but currently doesn't and trying to do it the same way as e.g. agetty results in sddm blocking the terminal itself and sddm-helper failing literally the same way.

the issue could have been avoided with e.g. this systemd unit (first one is what was packaged with sddm, 2nd is what I created to resolve this issue here. However there is some special handling that would be desirable for us that currenctly cant be used (`StandardInput=tty`, and `StandardOutput=tty` https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#StandardInput=).
> **StandardInput=**: (...)
> If null is selected, standard input will be connected to /dev/null, i.e. all read attempts by the process will result in immediate EOF.
>
> If tty is selected, standard input is connected to a TTY (as configured by TTYPath=, see below) and the executed process becomes the controlling process of the terminal. If the terminal is already being controlled by another process, the executed process waits until the current controlling process releases the terminal.
>
> tty-force is similar to tty, but the executed process is forcefully and immediately made the controlling process of the terminal, potentially removing previous controlling processes from the terminal.
>
> tty-fail is similar to tty, but if the terminal already has a controlling process start-up of the executed process fails.
> (...)
> This setting defaults to null, unless StandardInputText=/StandardInputData= are set, in which case it defaults to data.

Workaround that mostly works but requires knowing each and every other unit by name that is potentially using /dev/tty1.
```
[root@PC-001 ~]# systemctl cat sddm.service
# /usr/lib/systemd/system/sddm.service
[Unit]
Description=Simple Desktop Display Manager
Documentation=man:sddm(1) man:sddm.conf(5)
Conflicts=getty@tty1.service
After=systemd-user-sessions.service getty@tty1.service plymouth-quit.service systemd-logind.service
PartOf=graphical.target
StartLimitIntervalSec=30
StartLimitBurst=2

[Service]
ExecStart=/usr/bin/sddm
Restart=always

[Install]
Alias=display-manager.service


# /etc/systemd/system/sddm.service.d/override.conf
[Unit]
Conflicts=console-getty.service
After=console-getty.service
ConditionPathExists=/dev/tty1

[Service]
RestartSec=0
StandardInput=null
StandardOutput=journal
TTYPath=/dev/tty1
TTYReset=yes
TTYVHangup=yes
IgnoreSIGPIPE=no
SendSIGHUP=yes
ImportCredential=tty.tty1.agetty.*:agetty.
ImportCredential=tty.tty1.login.*:login.
ImportCredential=agetty.*
ImportCredential=login.*
ImportCredential=shell.*
```

What is needed to remediate this issue entirely is:
* sddm needs to be able to accept a tty device on stdin, stdout, and stderr (at a minimum at stdin for the special locking mechanism of systemd to take effect).
* sddm must handle receiving a SIGHUP which means it lost access to the tty to some other service unit.
* afterwards this sddm.service unit options should be used (doesn't need to know all other units by name that may be using /dev/tty1):
```
# /usr/lib/systemd/system/sddm.service
[Unit]
Description=Simple Desktop Display Manager
Documentation=man:sddm(1) man:sddm.conf(5)
After=systemd-user-sessions.service plymouth-quit.service systemd-logind.service getty-pre.target
Before=getty.target
PartOf=graphical.target
StartLimitIntervalSec=30
StartLimitBurst=2

ConditionPathExists=/dev/tty1

[Service]
ExecStart=/usr/bin/sddm
Restart=always
RestartSec=0
StandardInput=tty
StandardOutput=tty
UtmpIdentifier=tty1
TTYPath=/dev/tty1
TTYReset=yes
TTYVHangup=yes
IgnoreSIGPIPE=no
SendSIGHUP=yes
ImportCredential=tty.tty1.agetty.*:agetty.
ImportCredential=tty.tty1.login.*:login.
ImportCredential=agetty.*
ImportCredential=login.*
ImportCredential=shell.*

[Install]
WantedBy=getty.target
Alias=display-manager.service
```


While testing I didn't have xorg-server installed and had my `/etc/sddm.conf.d/settings.conf` set to:
```
[General]
DisplayServer=wayland
GreeterEnvironment=QT_WAYLAND_SHELL_INTEGRATION=layer-shell

[Wayland]
CompositorCommand=kwin_wayland --drm --no-lockscreen --no-global-shortcuts --locale1

[Theme]
EnableAvatars=false
```

**Comment by agowa** [comment]
Oh and it should also be noted that this specific log message does NOT show up when filtering for the sddm.service unit (e.g. `journalctl -xe --unit sddm.service`), but only when looking at all system logs. Despite almost all other messages from sddm-helper also being shown there.
`Jul 15 02:32:39 PC-001 sddm-helper[19054]: Failed to take control of "/dev/tty1" ("root"): Operation not permitted`

**Comment by Vogtinator** [comment]
> sddm needs to be able to accept a tty device on stdin, stdout, and stderr (at a minimum at stdin for the special locking mechanism of systemd to take effect).

The issue is that sddm has to manage ttys itself, only accepting a single tty is not sufficient. To fully leverage systemds support there would require having something like an `sddm-session@ttyX.service` template unit which would require design changes that effectively make systemd mandatory...

**Comment by davidedmundson** [comment]
> To fully leverage systemds support there would require having something like an `sddm-session@ttyX.service` template unit which would require design changes that effectively make systemd mandatory...

Replacing sddm-helper with a massive long systemd-run line is something that I think is very sensible and something I've played with. There's a service file in plasma-mobile that works well as a base.

Auth can be completely redone too, systemd can take care of all the pam_session operations. It provides a mechanism so that it'll still unlock keyrings even if the password was provided by another session. 

If we lean on it it unlocks a lot.

>design changes that effectively make systemd mandatory...

Ack, doing that in SDDM itself would cause an uproar with some upset people breaking things.
But if there was a certain fork where we're making a breaking change anyway where users could continue using SDDM that might be a good opportunity?

**Comment by agowa** [comment]
> > sddm needs to be able to accept a tty device on stdin, stdout, and stderr (at a minimum at stdin for the special locking mechanism of systemd to take effect).
> 
> The issue is that sddm has to manage ttys itself, only accepting a single tty is not sufficient.

Well it would be sufficient for solving this error. The reason is mainly the implicit "waiting for exclusive access" behavior that `StandardInput=tty` within the unit file provides. Currently doing so results in sddm's stdin getting attached to /dev/tty1 and the sddm-helper being unable to use it. It'd even be sufficient if there was a cli parameter for sddm that made it ignore stdin...

> To fully leverage systemds support there would require having something like an `sddm-session@ttyX.service` template unit which would require design changes that effectively make systemd mandatory...

true, that would be the best optimal usage of systemds capabilities and integrate best. But it is beyond the scope of this issue. I think this design change should be the next goal after, though. It would allow to avoid similar issues down the line. Alternatively someone could have a look at the code base of gdm as they for some reason manage to "not crash" right without relying upon systemd for locking tty1 for exclusive access. (their service unit is basically the same as sddm's right now, so there is probably also some alternative approach too)

> Replacing sddm-helper with a massive long systemd-run line is something that I think is very sensible and something I've played with. There's a service file in plasma-mobile that works well as a base.
> (...)
> Ack, doing that in SDDM itself would cause an uproar with some upset people breaking things.
> But if there was a certain fork where we're making a breaking change anyway where users could continue using SDDM that might be a good opportunity?

If as you say the base is already in plasma-mobile then I don't see a reason for a fork, add it as a new option and deprecate the old should there not be enough people willing to maintain it moving forward down the line...

**Comment by daniel-pg** [comment]
> On Artix + Dinit, disabled service tty1
> 
> and in /etc/dinit.d/config/console.conf set
> 
> `ACTIVE_CONSOLES="/dev/tty[2-6]" `
> 
> SDDM starts with `DisplayServer=wayland`

How do you disable the service?
Is `sudo dinitctl disable getty@tty1` enough? Also, editing /etc/dinit.d/config/console.conf completely breaks my system such that I can't even open another tty, and I had to fix it from a live CD.  

Edit: Never mind, I gave up on trying to make sddm work with Wayland, so I just use X11 to launch my Wayland session.
